### PR TITLE
FINERACT-1721: Delinquency classification as a Loan COB business step

### PR DESCRIPTION
**Pull request by josehernandezfintecheandomx**
## Description

Instead of only supporting the Delinquency Classification as a separate job, we need to make sure that it can be configured as a Business Step as part of the COB. 

There might be use-cases where in a Loan COB, another business step wants to build upon the delinquency information but serializing the Delinquency Classification job and the Loan COB steps is quite challenging so it’s best if the Delinquency Classification logic can be included as part of the Loan COB.

[FINERACT-1721](https://issues.apache.org/jira/browse/FINERACT-1721).

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java [hunk 1]
```diff
@@ -148,7 +148,7 @@ public CommandProcessingResult deleteDelinquencyBucket(Long delinquencyBucketId,
     public CommandProcessingResult applyDelinquencyTagToLoan(Long loanId, JsonCommand command) {
         Map<String, Object> changes = new HashMap<>();
 
-        final Loan loan = this.loanRepository.findOneWithNotFoundDetection(loanId);
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java]: As the loan is not used as a reference, rather doing actions on it, we should use the "this.loanRepository.findOneWithNotFoundDetection"

Using the "getReferenceById" is only useful if we dont need to fetch the loan entirely (fields and associations), but only we need the reference of it. 
Example:
`new LoanTransaction(loan, transactionDate,....)`

In the example we dont need to fetch the loan, only the reference which can be used to associate the loan entity with the loan transaction entity.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java [hunk 2]
```diff
@@ -166,7 +166,7 @@ public CommandProcessingResult applyDelinquencyTagToLoan(Long loanId, JsonComman
 
     @Override
     public void applyDelinquencyTagToLoan(Long loanId, Long ageDays) {
-        final Loan loan = this.loanRepository.findOneWithNotFoundDetection(loanId);
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java]: As the loan is not used as a reference, rather doing actions on it, we should use the "this.loanRepository.findOneWithNotFoundDetection"

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java [hunk 3]
```diff
@@ -329,7 +329,7 @@ private Map<String, Object> lookUpDelinquencyRange(final Loan loan, final Delinq
     }
 
     public Map<String, Object> setLoanDelinquencyTag(Long loanId, Long delinquencyRangeId) {
-        final Loan loan = this.loanRepository.findOneWithNotFoundDetection(loanId);
+        final Loan loan = this.loanRepository.getReferenceById(loanId);
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java]: Nice catch! We dont need to fetch the full loan here. Reference is enough! :)

**Review by adamsaghy — COMMENTED** [review]
@josehernandezfintecheandomx Please register this new Business step as part of the Loan COB Step.
For reference:
0022_add_batch_business_step_configuration_table.xml

Please write a test which execute Loan COB job and check whether the delinquency classification is applied propely on the overdue loan

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java (hunk 3 above)
Comment (adamsaghy) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java]: Would you please check whether this and the only method (public CommandProcessingResult setLoanDelinquencyTag(Long loanId, Long delinquencyRangeId, JsonCommand command)) in the line 108 is used at all?

We might not need this method at all...

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java (hunk 1 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java]: Done!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java (hunk 2 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java]: Done!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java (hunk 3 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java]: Done!

File: fineract-provider/src/main/java/org/apache/fineract/cob/loan/SetLoanDelinquencyTagsBusinessStep.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+@Slf4j
+@Component
+@RequiredArgsConstructor
+public class SetLoanDelinquencyTagsBusinessStep implements LoanCOBBusinessStep {
+
+    private final LoanAccountDomainService loanAccountDomainService;
+
+    @Override
+    public Loan execute(Loan loan) {
+        log.info("Set Loan Delinquency Tags Business Step {}", loan.getId());
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/cob/loan/SetLoanDelinquencyTagsBusinessStep.java]: Please do not log info level! Use debug!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java (hunk 1 above)
Comment (adamsaghy) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java]: @josehernandezfintecheandomx There is no change. You are still using referenceId

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java (hunk 2 above)
Comment (adamsaghy) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/delinquency/service/DelinquencyWritePlatformServiceImpl.java]: @josehernandezfintecheandomx no change... still using referenceId

File: fineract-provider/src/main/java/org/apache/fineract/cob/loan/LoanCOBWorkerConfiguration.java [hunk 5]
```diff
@@ -57,19 +60,42 @@ public class LoanCOBWorkerConfiguration {
...
+    }
+
     @Bean
     public Job loanCOBWorkerJob() {
         return jobBuilderFactory.get("Loan COB worker").start(loanCOBWorkerStep()).incrementer(new RunIdIncrementer()).build();
     }
 
+    @Bean
+    public Flow flow() {
+        return new FlowBuilder<Flow>("cobFlow").start(initialisationStep()).next(loanBusinessStep()).next(loanBusinessStep()).build();
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/cob/loan/LoanCOBWorkerConfiguration.java]: Duplicate ".next(loanBusinessStep())". Please remove one of them

Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/cob/loan/LoanCOBWorkerConfiguration.java]: Duplicate ".next(loanBusinessStep())". Please remove one of them

**Review by adamsaghy — APPROVED** [review]
LGTM

**Review by galovics — APPROVED** [review]

**Pull request merged**